app/src/main/java/com/epfl/drawyourpath/mainpage/fragments/ChatFragment.kt

```diff
@@ -1,43 +1,251 @@
+                        // Find the selected friend's ID
+                        val selectedFriendId = userData.friendList?.firstOrNull { friendId ->
+                            try {
+                                database.getUsername(friendId).get() == selectedFriendUsername
+                            } catch (e: Exception) {
+                                Log.e(TAG, "Error while getting friend username: ", e)
+                                false
+                            }
+                        }
+                        if (selectedFriendId != null) {
+                            // If a friend is selected, create a new chat conversation
+                            val chatName = "Chat with $selectedFriendUsername"
+                            val membersList = listOf(userId, selectedFriendId)
+                            val welcomeMessage = getString(R.string.chat_welcome_message)
+                            database.createChatConversation(
+                                chatName,
+                                membersList,
+                                userId,
+                                welcomeMessage,
+                            )
+                                .thenAccept { conversationId ->
+                                    Log.d(
+                                        "ChatCreation",
+                                        "Chat created with conversationId: $conversationId",
+                                    )
+
+                                    // Create a new ChatPreview object for the newly created chat
+                                    val welcomeMessageContent =
+                                        MessageContent.Text(getString(R.string.chat_welcome_message)) // Replace with your MessageContent creation
+                                    val welcomeMessage = Message(
+                                        0L,
+                                        userId,
+                                        welcomeMessageContent,
+                                        System.currentTimeMillis(),
+                                    )
+                                    val newChatPreview =
+                                        ChatPreview(conversationId, chatName, welcomeMessage)
+
+                                    // Add the new ChatPreview to the chatList and notify the adapter
+                                    requireActivity().runOnUiThread {
+                                        chatList.add(newChatPreview)
+                                        chatAdapter.notifyDataSetChanged()
+                                    }
+                                }
+                        }
+                    }
+                    true
+                }
+                popupMenu.show()
+            }.exceptionally { exception ->
+                Log.e(TAG, "Error while getting UserAccount: ", exception)
+                null
+            }
+        }
+    }

```

Comment:
Maybe you can create some helpers functions to have a better structure of the code, since it's very difficult to understand this block of code.

Reply:
I think ill have it as a small task for the next spring as it's not a trivial thing to do. I added a TODO.